### DISCUSSION TITLE
how to use pcf8574 IIC extended GPIO for pushbutton and relay

### DISCUSSION THREAD

**Discussion by hzkincony** [HomeSpan/HomeSpan#394] [Q&A]
i have used your RealPushBotton demo code, it work well with relay and push bottons. this use ESP32 GPIO pin number directly. But now i have used PCF8574 IIC extend for get more GPIOs, how to define many GPIOs for relay and push buttons? i don't know which file to edit. Thanks!

**Comment by HomeSpan** [comment]
What library do you use now for your PCF8574?  You should include that in your HomeSpan sketch and access the pins exactly the same way you would if you were not using HomeSpan.  For outputs, there is nothing specific to HomeSpan - simply use whatever command you were already using for the PCF8574 for sketches not involving HomeSpan.  For inputs, you can configure `SpanButton()` to use whatever command you normally use to read from the PCF8574.  Exact instructions on how to do this are shown at the end of the RealPushButton demo code.

**Reply by hzkincony** [reply]
thanks,  i use this PCF8574 library: https://www.mischianti.org/2019/01/02/pcf8574-i2c-digital-i-o-expander-fast-easy-usage/
i can include my PCF8574 library. But in your demo code:
such as:
new DEV_DimmableLED(2,36);    //relay pin, push botton pin

but my pcf8574 extend GPIO such as:

pcf8574.pinMode(P1, OUTPUT);
pcf8574.pinMode(P2, INPUT);

how do i replace these: (2,36) ?

**Comment by hzkincony** [comment]
hello, can you help me?

**Comment by HomeSpan** [comment]
The parameters (2,36) in the structure `DEV_DimmableLED()` are simply pass-through variables that the logic inside `DEV_DimmableLED()` uses.  As with any C++ structure, you can define any variables you want, as many as you want, and they can mean anything you'd like.   The only thing that matters is how you interpret them within the logic inside the structure.  You don't even have to specify any variables and can instead hardcode pin numbers directly in the structure.

For example, if you wanted to pass a relay pin number and a button pin number to the structure, you can define it as such:

```
DEV_DimmableLED(int relayPin, int buttonPin) : Service::LightBulb() 
```

What you do with these variables depends on the specifics of your device.  As you can see in Example 5, to turn on an LED directly attached to a pin, I use `digitalWrite(ledPin,power->getNewVal());`.  If you are using a GPIO extender, that won't work.  Instead, you'll need to replace `digitalWrite()` with whatever your library command is for controlling a pin on the GPIO extender.  Similarly, instead of initializing the pin with `pinMode(ledPin,OUTPUT)` you'll need to use your library command, which based on the above should be `pcf8574.pinMode(relayPin, OUTPUT)`.

The same logic applied to using a GPIO extender with a SpanButton but there are a few extra steps that need to be taken.  I recommend you first try implementing just the relay portion of the project with the GPIO extender - when that's all working let me know and we can see about adding the SpanButton code.

**Reply by rubensx** [reply]
I'm looking for this SpanButton code for the pcf8574 inputs. I'm trying programmable buttons but using example 16 ProgrammableSwitchesbut I'm stucking here:

    I'm trying some alternatives but can't figure out how I can make these buttons to work, Can you spot some light about the DEV_pushbutton.h can be mod to see the i2s inputs?


```
#include "PCF8574.h"
PCF8574 pcf8574_IN1(0x22,4,5);

pcf8574_IN1.pinMode(P0, INPUT);

int val1 = pcf8574_IN1.digitalRead(P0);

  homeSpan.begin(Category::Bridges,"HomeSpan Bridge");
  
  
new SpanAccessory();                                                          
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      new Characteristic::Name("PushButton Switches");

    // We've written DEV_ProgButton to take two arguments.  The first is a pin number that DEV_ProgButton.h uses to create a SpanButton.  The second is an index number
    // that HomeKit uses as a label when you program the actions of each button in the Home App.  The numbers do not have to be sequential, nor start with 1.  They just need
    // to be unique so HomeKit can distinguish them.  Note that HomeKit does not require index numbers if you only have one Stateless Programmable Switch Service within any
    // given Accessory.  Since we have two, we must specify two unique index numbers.
    
    new DEV_ProgButton(val1,1);       // trying to put "val1" here but the result is pin=0 in homespan feddback
```

**Reply by HomeSpan** [reply]
I recommend starting first with Example 15 --- the comments at the end explain exactly how to use custom functions in SpanButton.

Unless you modified the code for DEV_ProgButton() above, this can't work.  Note in the Programmable Button example that the first argument to DEV_ProgButton() is an integer, which represents the **pin number** to which a pushbutton is attached.  You are passing it a specific read value, which you set at the top of your code as the return value from a single reading of P0 from the pcf8574.  Instead, you should be passing `P0` as the first argument and the line `pcf8574_IN1.digitalRead()` should be used from within DEV_ProgButton() to perform the actual reading.

**Reply by rubensx** [reply]
Yeah I try to use:

 new DEV_ProgButton(pcf8574_IN1.digitalRead(P0),1);

but when I do that, the ESP32 stops to charge in Homespan loading serial output and freeze! 

Did you figure out how I can make this sentece work?

**Reply by HomeSpan** [reply]
Sorry, that's not what I meant by using `pcf8574` from within DEV_ProgButton().

Note that `DEV_ProgButton()` is *not* a function.  It does not "do" anything.  Rather it is a structure/class that contains a bunch of variables and different functions that implement the actions you would like to take via different HomeSpan protocols.  You need to modify the code *inside* that class to incorporate the pcf8574 logic you want by writing/modifying the own code based on the pcf8574 to perform the functions your need (such as reading a GPIO Extender pin and changing the state of a HomeSpan Characteristic).   Within that structure you can also instantiate SpanButton using your own logic as described in Example 15.  

If you've not used C++ before, or are unfamiliar with structures and class, and how they work in HomeSpan, you may want to review the first few Tutorial Examples as well as the HomeSpan Overview page that provides detailed instructions.

**Reply by rubensx** [reply]
oh no... I'm still not a great programmer to do that! I don't know if I can post another library here but, later (without a kincony board) I made by myself a board and use another library  from:

[https://github.com/Mixiaoxiao/Arduino-HomeKit-ESP32/blob/master/examples/ESP32_homekit_wolfssl_4_3_0/ESP32_homekit.ino](url)

and all work ok in my 8 chanel relays and 8 input switches directly in esp32 but I had a problem with a index label of switches and migrate to homespan to solve this...

it's not very easy to understand but maybe this feature to expand inputs and outputs are desirable!

I'll try other ways but if you mind to create a support to pcf8574 i2c library or something else I'm here to test and contribute

**Reply by rubensx** [reply]
Any chances to modify and change the interpretation of the "PIN" as a Alias "or state of"? 
can you show me the exactly part on a Homespan Library who maybe I can do some adaptation from PCF8574 library to work?

**Reply by HomeSpan** [reply]
There is nothing you need to do to modify the library itself -- you just have to use the existing API functions to code your sketch to work with the PCF8574 library. 

If you can provide some details about exactly what you are trying to do, I may be able to provide some sample code.

**Reply by rubensx** [reply]
I need to create a Programmable switch Buttons instead a linked button, but as I mention before, using your 16 example, I need to put a Pin in this sentence:

DEV_ProgButton(12,1);
where 12 is the pin and 1 a index, ok!

but when I use PCF8574.h, I need to use an example:
#define Pin1 pcf8574_IN1.digitalRead(P0) 

(Where P0 from pcf8574 act as input and pass the readin to “Pin1”)

and the sentence use to be:
DEV_ProgButton(Pin1,1);

but, homespan don’t recognize the Pin1 as a Pin and goes to 0 in output serial

I’ve loosing something…

**Reply by rubensx** [reply]
When I put this variable the system stuck in start:

```
ets Jul 29 2019 12:21:46

rst:0x1 (POWERON_RESET),boot:0x17 (Sets Jul 29 2019 12:21:46

rst:0x1 (POWERON_RESET),boot:0x17 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13260
load:0x40080400,len:3028
entry 0x400805e4
Init pcf8574_IN1... success!
Init pcf8574_RE... success!

************************************************************
Welcome to HomeSpan!
Apple HomeKit for the Espressif ESP-32 WROOM and Arduino IDE
************************************************************

** Please ensure serial monitor is set to transmit <newlines>

Message Logs:     Level 2
Status LED:       Pin -  *** WARNING: Status LED Pin is UNDEFINED
Device Control:   Pin -  *** WARNING: Device Control Pin is UNDEFINED
Sketch Version:   n/a
HomeSpan Version: 1.8.0
Arduino-ESP Ver.: 2.0.11
ESP-IDF Version:  4.4.5
ESP32 Chip:       ESP32-D0WD-V3 Rev 3 dual-core 4MB Flash
ESP32 Board:      doitESP32devkitV1
PWM Resources:    16 channels, 8 timers, max 20-bit duty resolution
Sodium Version:   1.0.12-idf  Lib 9.4
MbedTLS Version:  mbed TLS 2.28.3
Sketch Compiled:  Sep  4 2023 15:20:59
Partition:        app0
MAC Address:      B0:B2:1C:CE:D5:94

Device Name:      HomeSpan Bridge
```

**Reply by HomeSpan** [reply]
@rubensx, I don't have a PCF8574 to test, but using the library you mentioned above I modified Example 16 to use the PCF8574 for a SpanButton.  The code below is an exact copy of Example 16 except that I've removed all the original comments and added new comments wherever I modified the code.  Please give it a try - as mentioned I cannot test it, but it does compile correctly.  You may have to perform your own debugging as I won't be able to do so.

Here is the revised main *.ino file:

```C++
#include "HomeSpan.h" 

#include "PCF8574.h"        // INCLUDE THE LIBRARY
PCF8574 pcf8574(0x39);      // CREATE THE OBJECT USING WHATEVER ADDRESS YOU SET FOR THE EXTENDER

#include "DEV_ProgButton.h"     

void setup() {
  
  Serial.begin(115200);

  homeSpan.begin(Category::Bridges,"HomeSpan Bridge");

  new SpanAccessory();  
    new Service::AccessoryInformation();
      new Characteristic::Identify();

  new SpanAccessory();                                                          
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      new Characteristic::Name("PushButton Switches");

  // THE PIN NUMBERS BELOW REFER TO THE PINS ON THE PCF8574 
  
    new DEV_ProgButton(5,1);       // create Stateless Programmable Switch Service on PCF8574 pin 5 with index=1
    new DEV_ProgButton(6,2);       // create Stateless Programmable Switch Service on PCF8574 pin 6 with index=2

    pcf8574.begin();               // START THE PCF8574
    
} // end of setup()

//////////////////////////////////////

void loop(){
  
  homeSpan.poll();
  
} // end of loop()
```

Here is the revised DEV_ProgButton.h file:

```C++

////////////////////////////////////////////////////
//   DEVICE-SPECIFIC PROGRAMMABLE SWITCH SERVICES //
////////////////////////////////////////////////////

struct DEV_ProgButton : Service::StatelessProgrammableSwitch {
 
  SpanCharacteristic *switchEvent;
  
  DEV_ProgButton(int buttonPin, int index) : Service::StatelessProgrammableSwitch(){

    switchEvent=new Characteristic::ProgrammableSwitchEvent();
    new Characteristic::ServiceLabelIndex(index);

    pcf8574.pinMode(buttonPin, INPUT_PULLUP);     // SET THE PIN ON THE PCF8574 TO INPUT WITH PULLUP
    
    new SpanButton(buttonPin,PCF_READ);           // CREATE A SPANBUTTON, BUT HAVE SPANBUTTON CALL A CUSTOM FUNCTION (DEFINED BELOW) TO READ THE PCF8574

    Serial.print("Configuring Programmable Pushbutton: Pin=");
    Serial.print(buttonPin);
    Serial.print("  Index=");
    Serial.print(index);
    Serial.print("\n");
    
  }

  void button(int pin, int pressType) override {

    LOG1("Found button press on pin: ");
    LOG1(pin);
    LOG1("  type: ");
    LOG1(pressType==SpanButton::LONG?"LONG":(pressType==SpanButton::SINGLE)?"SINGLE":"DOUBLE");
    LOG1("\n");

    switchEvent->setVal(pressType);
  }

  static boolean PCF_READ(int pcfPin) {         // THIS IS A CUSTOM FUNCTION THAT RETURN THE VALUE OF A PCF PIN USING THE PCF8574 LIBRARY
    return(!pcf8574.digitalRead(pcfPin));       // THIS FUNCTION ASSUMES YOU HAVE A PUSHBUTTON CONNECTING THE PCF PIN TO *GROUND*
  }  

};
      
//////////////////////////////////
```

**Reply by rubensx** [reply]
Thank you for this help!

first "main code" has problems:

When you create this:

```
new DEV_ProgButton(5,1);       // create Stateless Programmable Switch Service on PCF8574 pin 5 with index=1 
new DEV_ProgButton(6,2);       // create Stateless Programmable Switch Service on PCF8574 pin 6 with index=2
```
    
    
You need to define the PCF8574 ports to "call"  this port as 5 or whatever you want... Like you describe before, the ESP32 will looking for the 5 and 6 logical port and not I2C PCF8574 ports...

So, you need to name each pcf port like this:

```
    int D1 = pcf8574.digitalRead(P0);
    int D2 = pcf8574.digitalRead(P1);
```
    
Need to define these ports too:

```
pcf8574.pinMode(P0, INPUT);
pcf8574.pinMode(P1, INPUT);
pcf8574.pinMode(P2, INPUT);
pcf8574.pinMode(P3, INPUT);
pcf8574.pinMode(P4, INPUT);
pcf8574.pinMode(P5, INPUT);
pcf8574.pinMode(P6, INPUT);
pcf8574.pinMode(P7, INPUT);
```

and each port is called "P0", "P1"....

Ok... with "D1" or "D2" named, the result in serial is  Pin=0 :

```
Device Name:      HomeSpan Bridge

Configuring Programmable Pushbutton: Pin=0  Index=1
Configuring Programmable Pushbutton: Pin=0  Index=2

```
If I put the P0 and P1 directly in sentence

```
new DEV_ProgButton(P0,1);       // create Stateless Programmable Switch Service on PCF8574 pin 5 with index=1 
new DEV_ProgButton(P1,2);       // create Stateless Programmable Switch Service on PCF8574 pin 6 with index=2
```

The Result:

```
Device Name:      HomeSpan Bridge

Configuring Programmable Pushbutton: Pin=0  Index=1
Configuring Programmable Pushbutton: Pin=1  Index=2

```
Now, he "see" the number but not the P... and do nothing if I push the buttons hooked in P0 or P1 port (and nothing directly the 0 pin port of ESP32)

Maybe we need to create a variable to pass the state to the button?

**Reply by rubensx** [reply]
oh I undestand later you create a "function"

```
static boolean PCF_READ(int pcfPin) {         // THIS IS A CUSTOM FUNCTION THAT RETURN THE VALUE OF A PCF PIN USING THE PCF8574 LIBRARY
    return(!pcf8574.digitalRead(pcfPin));       // THIS FUNCTION ASSUMES YOU HAVE A PUSHBUTTON CONNECTING THE PCF PIN TO *GROUND*
```
    
    
  and don't need my whole name D2 or D1, but in your code you need to put a "P" before de pcfPin, but still the "trigger" is not recognized by Spanbutton, only a number...

**Reply by HomeSpan** [reply]
It seems you are very new to programming - please note that using HomeSpan requires a good deal of programming.  It is not designed to be plug-and-play, but requires a good deal of code writing to use successfully.

Based on the specifications for the PCF8574 library provided by https://www.mischianti.org/2019/01/02/pcf8574-i2c-digital-i-o-expander-fast-easy-usage/, the pins used on the PCF are **numbered 0-7**.   In most programming languages the maker of the library may provide names for such things.  In this case the author, mischianti,  defined P0=0, P1=1, P2=2, so that in the code you can use EITHER the number 4 or the symbol P4.  When the code is compiled, every instance of the symbol P4 is replaced by the number 4.  That's why when you say P0 and P1 in the above code, HomeSpan shows 0 and 1.  There is no P0 pin.  P0 is just a reference to the number 0 itself.

Also, the line `int D1 = pcf8574.digitalRead(P0);` does not at all "name" a port.  That's not how programming in C++ works.  Rather, in this line you created a variable D1, which is not needed anywhere, and then you set its value to either by 0 or 1 by reading pin P0.  There is nothing wrong with this line - but it does nothing useful for what you are trying to accomplish.

I'm not sure there is much more I can offer in the way of help.  I don't know how your configured your hardware, and as mentioned I don't have a PCF8574 chip to test out on my own.

**Reply by rubensx** [reply]
Yes I’m not a professional programmer (a hobbist really) but, running your code as is, doesn’t work saddly 

the library is the same and I’m using a Kincony KC868-A8S hardware

https://www.kincony.com/kc868-a8s-hardware-design-details.html

The relays are running ok in homespan but the inputs… 

I’m trying to figure out how to emulate a input door to match with homespan… This I2C are interesting to expand possibilities at all but I guess we need a specific library to match each with other

**Reply by rubensx** [reply]
Hey I'm persistent I know, but...

Maybe you can help to investigate that, I do some edit in your code and now, some progress (strange progress) , please look that


```
#include "HomeSpan.h" 

#include "PCF8574.h"        // INCLUDE THE LIBRARY
PCF8574 pcf8574(0x22,4,5);      // CREATE THE OBJECT USING WHATEVER ADDRESS YOU SET FOR THE EXTENDER

#include "DEV_ProgButton.h"     

void setup() {
  
  Serial.begin(115200);
 homeSpan.setLogLevel(2);

  pcf8574.pinMode(0, INPUT_PULLUP);
  pcf8574.pinMode(1, INPUT_PULLUP);
  pcf8574.pinMode(2, INPUT_PULLUP);
  pcf8574.pinMode(3, INPUT_PULLUP);
  pcf8574.pinMode(4, INPUT_PULLUP);
  pcf8574.pinMode(5, INPUT_PULLUP);
  pcf8574.pinMode(6, INPUT_PULLUP);
  pcf8574.pinMode(7, INPUT_PULLUP);

    pcf8574.begin();               // START THE PCF8574




  homeSpan.begin(Category::Bridges,"HomeSpan Bridge");

  new SpanAccessory();  
    new Service::AccessoryInformation();
      new Characteristic::Identify();

  new SpanAccessory();                                                          
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      new Characteristic::Name("PushButton Switches");
      

  // THE PIN NUMBERS BELOW REFER TO THE PINS ON THE PCF8574 
  
    new DEV_ProgButton(0,1);       // create Stateless Programmable Switch Service on PCF8574 pin 0 with index=1
    new DEV_ProgButton(1,2);       // create Stateless Programmable Switch Service on PCF8574 pin 1 with index=2
    new DEV_ProgButton(2,3);       // create Stateless Programmable Switch Service on PCF8574 pin 2 with index=3
    new DEV_ProgButton(3,4);       // create Stateless Programmable Switch Service on PCF8574 pin 3 with index=4
 

    
} // end of setup()

//////////////////////////////////////

void loop(){

    
  uint8_t val0 = pcf8574.digitalRead(0);
  if (val0==LOW) Serial.println("KEY 0 PRESSED");
  delay(50);
  uint8_t val1 = pcf8574.digitalRead(1);
  if (val1==LOW) Serial.println("KEY 1 PRESSED");
  delay(50);
   uint8_t val2 = pcf8574.digitalRead(2);
  if (val2==LOW) Serial.println("KEY 2 PRESSED");
  delay(50);
  uint8_t val3 = pcf8574.digitalRead(3);
  if (val3==LOW) Serial.println("KEY 3 PRESSED");
  delay(50);
   uint8_t val4 = pcf8574.digitalRead(4);
  if (val4==LOW) Serial.println("KEY 4 PRESSED");
  delay(50);
  uint8_t val5 = pcf8574.digitalRead(5);
  if (val5==LOW) Serial.println("KEY 5 PRESSED");
  delay(50);
   uint8_t val6 = pcf8574.digitalRead(6);
  if (val6==LOW) Serial.println("KEY 6 PRESSED");
  delay(50);
  uint8_t val7 = pcf8574.digitalRead(7);
  if (val7==LOW) Serial.println("KEY 7 PRESSED");
  delay(50);
  
  homeSpan.poll();
  
} // end of loop()
```

and

```
///////////////////////////////////////////////////
//   DEVICE-SPECIFIC PROGRAMMABLE SWITCH SERVICES //
////////////////////////////////////////////////////


struct DEV_ProgButton : Service::StatelessProgrammableSwitch {
 
  SpanCharacteristic *switchEvent;
  
  DEV_ProgButton(int buttonPin, int index) : Service::StatelessProgrammableSwitch(){

    switchEvent=new Characteristic::ProgrammableSwitchEvent();
    new Characteristic::ServiceLabelIndex(index);

   
    new SpanButton(buttonPin,PCF_READ);           // CREATE A SPANBUTTON, BUT HAVE SPANBUTTON CALL A CUSTOM FUNCTION (DEFINED BELOW) TO READ THE PCF8574


    Serial.print("Configuring Programmable Pushbutton: Pin=");
    Serial.print(buttonPin);
    Serial.print("  Index=");
    Serial.print(index);
    Serial.print("\n");
    
  }

  void button(int pin, int pressType) override {

    LOG1("Found button press on pin: ");
    LOG1(pin);
    LOG1("  type: ");
    LOG1(pressType==SpanButton::LONG?"LONG":(pressType==SpanButton::SINGLE)?"SINGLE":"DOUBLE");
    LOG1("\n");

    switchEvent->setVal(pressType);
  }

  static boolean PCF_READ(int pcfPin) {         // THIS IS A CUSTOM FUNCTION THAT RETURN THE VALUE OF A PCF PIN USING THE PCF8574 LIBRARY
    return(!pcf8574.digitalRead(pcfPin));       // THIS FUNCTION ASSUMES YOU HAVE A PUSHBUTTON CONNECTING THE PCF PIN TO *GROUND*
  }  

};
      
//////////////////////////////////
```


Ok Let's to the strange part: When I put in loop the line

```
 uint8_t val0 = pcf8574.digitalRead(0);
  if (val0==LOW) Serial.println("KEY 0 PRESSED");
  delay(50);
```

The Homespan respond in Serial:

```
ets Jul 29 2019 12:21:46

rst:0x1 (POWERON_RESET),boot:0x17 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13260
load:0x40080400,len:3028
entry 0x400805e4

************************************************************
Welcome to HomeSpan!
Apple HomeKit for the Espressif ESP-32 WROOM and Arduino IDE
************************************************************

** Please ensure serial monitor is set to transmit <newlines>

Message Logs:     Level 2
Status LED:       Pin -  *** WARNING: Status LED Pin is UNDEFINED
Device Control:   Pin -  *** WARNING: Device Control Pin is UNDEFINED
Sketch Version:   n/a
HomeSpan Version: 1.8.0
Arduino-ESP Ver.: 2.0.11
ESP-IDF Version:  4.4.5
ESP32 Chip:       ESP32-D0WD-V3 Rev 3 dual-core 4MB Flash
ESP32 Board:      doitESP32devkitV1
PWM Resources:    16 channels, 8 timers, max 20-bit duty resolution
Sodium Version:   1.0.12-idf  Lib 9.4
MbedTLS Version:  mbed TLS 2.28.3
Sketch Compiled:  Sep  5 2023 17:20:13
Partition:        app0
MAC Address:      B0:B2:1C:CE:D5:94

Device Name:      HomeSpan Bridge

Configuring Programmable Pushbutton: Pin=0  Index=1
Configuring Programmable Pushbutton: Pin=1  Index=2
Configuring Programmable Pushbutton: Pin=2  Index=3
Configuring Programmable Pushbutton: Pin=3  Index=4

*** HomeSpan Info ***

‚û§ Accessory:  AID=1
   ‚ûü Service AccessoryInformation:  IID=1, UUID="3E"
      ‚á® Characteristic Identify(0):  IID=2, UUID="14", Perms=PW
‚û§ Accessory:  AID=2
   ‚ûü Service AccessoryInformation:  IID=1, UUID="3E"
      ‚á® Characteristic Identify(0):  IID=2, UUID="14", Perms=PW
      ‚á® Characteristic Name("PushButton Switches"):  IID=3, UUID="23", Perms=PR
   ‚ûü Service StatelessProgrammableSwitch:  IID=4, UUID="89"
      ‚á® Characteristic ProgrammableSwitchEvent(0):  IID=5, UUID="73", Perms=PR+EV, Range=[0,2]
      ‚á® Characteristic ServiceLabelIndex(1):  IID=6, UUID="CB", Perms=PR, Range=[1,255]
      ‚ñº SpanButton: Pin=0, Single=5ms, Double=200ms, Long=2000ms, Type=USER-DEFINED
   ‚ûü Service StatelessProgrammableSwitch:  IID=7, UUID="89"
      ‚á® Characteristic ProgrammableSwitchEvent(0):  IID=8, UUID="73", Perms=PR+EV, Range=[0,2]
      ‚á® Characteristic ServiceLabelIndex(2):  IID=9, UUID="CB", Perms=PR, Range=[1,255]
      ‚ñº SpanButton: Pin=1, Single=5ms, Double=200ms, Long=2000ms, Type=USER-DEFINED
   ‚ûü Service StatelessProgrammableSwitch:  IID=10, UUID="89"
      ‚á® Characteristic ProgrammableSwitchEvent(0):  IID=11, UUID="73", Perms=PR+EV, Range=[0,2]
      ‚á® Characteristic ServiceLabelIndex(3):  IID=12, UUID="CB", Perms=PR, Range=[1,255]
      ‚ñº SpanButton: Pin=2, Single=5ms, Double=200ms, Long=2000ms, Type=USER-DEFINED
   ‚ûü Service StatelessProgrammableSwitch:  IID=13, UUID="89"
      ‚á® Characteristic ProgrammableSwitchEvent(0):  IID=14, UUID="73", Perms=PR+EV, Range=[0,2]
      ‚á® Characteristic ServiceLabelIndex(4):  IID=15, UUID="CB", Perms=PR, Range=[1,255]
      ‚ñº SpanButton: Pin=3, Single=5ms, Double=200ms, Long=2000ms, Type=USER-DEFINED

Configured as Bridge: YES

Database Validation:  Warnings=0, Errors=0

Service                             UUID         AID  IID  Update  Loop  Button  Linked Services
------------------------------  --------  ----------  ---  ------  ----  ------  ---------------
AccessoryInformation                  3E           1    1      NO    NO      NO  -
AccessoryInformation                  3E           2    1      NO    NO      NO  -
StatelessProgrammableSwitch           89           2    4      NO    NO     YES  -
StatelessProgrammableSwitch           89           2    7      NO    NO     YES  -
StatelessProgrammableSwitch           89           2   10      NO    NO     YES  -
StatelessProgrammableSwitch           89           2   13      NO    NO     YES  -

*** End Info ***

Accessory ID:      EB:4C:63:34:21:2E                               LTPK: A8E1D8D522F3AE36FA00978F32210F932D42C54B12C2BB9002279784B941CDB6
No Paired Controllers

Accessory configuration number: 1

*** WIFI CREDENTIALS DATA NOT FOUND.  YOU MAY CONFIGURE BY TYPING 'W <RETURN>'.

HomeSpan Bridge is READY!

KEY 0 PRESSED
Found button press on pin: 0  type: SINGLE
KEY 1 PRESSED
Found button press on pin: 1  type: SINGLE
Found button press on pin: 2  type: SINGLE
KEY 3 PRESSED
Found button press on pin: 3  type: SINGLE
KEY 4 PRESSED
KEY 5 PRESSED
KEY 5 PRESSED
KEY 6 PRESSED
KEY 6 PRESSED
KEY 6 PRESSED
KEY 7 PRESSED
KEY 7 PRESSED

```
I don't know exactly what happens here but finally I get a "Found button press on pin: 0  type: SINGLE" but I tottaly unknow why ...

If I suprime the part of the line:
 if (val0==LOW) Serial.println("KEY 0 PRESSED");
 
 Homespan can't identify any button press... 
 
 I'm sorry insistence about this but maybe you can help me understand why this happening

**Reply by HomeSpan** [reply]
I downloaded the specification for the pcf8574 and I'm not sure the device will work well with SpanButton.  Since it uses I2C there are delays in reading the data.  Also, the chip actually reports back all values P0-P7 every time it is read - the library is translating this into individual reads from a pin.  I'm not sure what the net effect of all this is on SpanButton.

To solve this I would need to experiment directly with the chip.  It may be that the most efficient way of reading is NOT to use SpanButton but simply read the data in a loop.  A lot of this depends on how sensitive the chip is to debouncing.  It's impossible for me to guess unless I spent a few days experimenting.  

I do not know why adding reads into the loop() as you did above allows SpanButton to detect some of the presses.  My guess is that it's all related to the length of time it take the device to process I2C commands.  

Feel free to experiment by changing the delay parameters in SpanButton - it may be that you need to make the SINGLE press delay longer, or maybe shorter.  As I said, it's all a matter of experimenting.  

You can change the delays in SpanButton by simply adding 3 more arguments like this:

`new SpanButton(buttonPin,PCF_READ,2000,5,200)` where the parameters are fully described in the [API page](https://github.com/HomeSpan/HomeSpan/blob/master/docs/Reference.md#spanbuttonint-pin-uint16_t-longtime-uint16_t-singletime-uint16_t-doubletime-boolean-triggertypeint).

Good luck experimenting and let us know if you find a combination that seems to work.

**Reply by rubensx** [reply]
In my tests I change the delays directly on the library (thank you about the new line data, I’m try this but get an error).

Really a bug between the I2C readings and a Homespan library but seems the simples change about these delays solve for a while! 

Is not a elegant solution but working good with delay(6) into a loop for a single press 60, double 400 at last long 2000

I like so much homespan and your work, I’ll implement some MLX90614 I2C temperature sensor to learn more possibilities! In this board I can test other things too

thank you for now and please call me here if you had another idea or tests!

cheers!

**Comment by hzkincony** [comment]
thank you very much, it can work now!
